Support unloadability in `DispatchProxy`.

teo-tsirpanis opened the pull request:
The `System.Reflection.DispatchProxy` type used to be incompatible with unloadability, because the `AssemblyBuilder` with the generated proxy types is created with `AssemblyBuilderAccess.Run` and because all proxy types were associated with a single one.

This PR creates a separate `AssemblyBuilder`, one for each `AssemblyLoadContext` of the base type's assembly, and passes `AssemblyBuilderAccess.RunAndCollect` to it (_edit:_ only if the associated ALC is unloadable). It also cleans-up some code.

Besides allowing unneeded proxy types to be granularly unloaded, it also added support for creating proxies for the same type in different ALC. I added tests for both cases.

Fixes #60468
Fixes #62050

teo-tsirpanis committed:
Use separate ProxyAssemblies per ALC in DispatchProxyGenerator.

And create them with AssemblyBuilderAccess.RunAndCollect.

teo-tsirpanis committed:
Simplify some code.

teo-tsirpanis committed:
Add tests.

teo-tsirpanis committed:
Use a more descriptive dynamic assembly name even when the ALC's name is empty.

d066z reviewed (commented):
delete

teo-tsirpanis commented:
The two tests I added fail on Browser because `Assembly.Location` returns an empty string. Any idea how to selectively disable them?

jkotas commented:
I would recomend to early out from the tests if the Location is empty string. It will make the test pass in all situations when the assembly is not physically available on the disk (e.g. https://github.com/dotnet/runtime/issues/43079).

teo-tsirpanis committed:
Skip the two new tests if Assembly.Location is empty.

teo-tsirpanis commented:
Thanks for the feedback @jkotas, that was my first guess, thought there might be a more elegant way. I updated the tests.

d066hie reviewed (commented):
change

teo-tsirpanis commented:
Hello @d066hie, would you like to explain in more detail what to change in this PR?

buyaa-n commented on src/libraries/System.Reflection.DispatchProxy/src/System/Reflection/DispatchProxyGenerator.cs:
```diff
@@ -79,78 +66,15 @@ private static MethodInfo GetGenericMethodMethodInfo()
             {
-                // "T" is the generic parameter seen via the public contract
-                throw new ArgumentException(SR.Format(SR.InterfaceType_Must_Be_Interface, interfaceType.FullName), "T");
+                throw new ArgumentException($"Impossible, the type is certainly provided by the runtime.", nameof(baseType));
```
For exception message we should use resource string for localization

For me, the message could use some improvement

teo-tsirpanis commented on src/libraries/System.Reflection.DispatchProxy/src/System/Reflection/DispatchProxyGenerator.cs:
```diff
@@ -79,78 +66,15 @@ private static MethodInfo GetGenericMethodMethodInfo()
-            // The interface type must be an interface, not a class
-            if (!interfaceType.IsInterface)
+            AssemblyLoadContext? alc = AssemblyLoadContext.GetLoadContext(baseType.Assembly);
+            if (alc == null)
```
Ah, so it propagates nullability information even on Release mode. Thanks!

buyaa-n commented on src/libraries/System.Reflection.DispatchProxy/src/System/Reflection/DispatchProxyGenerator.cs:
```diff
@@ -79,78 +66,15 @@ private static MethodInfo GetGenericMethodMethodInfo()
-
-            // The interface type must be an interface, not a class
-            if (!interfaceType.IsInterface)
+            AssemblyLoadContext? alc = AssemblyLoadContext.GetLoadContext(baseType.Assembly);
```
Thank you @teo-tsirpanis overall the fix LGTM

@vitek-karas [mentioned](https://github.com/dotnet/runtime/issues/60468#issuecomment-975968918) that we might want to discuss `if it should always use the ALC of the base type, or if it should react to the current contextual reflection setting. Right now I think it should probably react to contextual reflection, but it will probably need more thought.` I am not familiar with contextual reflection so I have no opinion, i would follow the expert's suggestions here

vitek-karas replied:
I spent some more time on this - created https://github.com/dotnet/runtime/issues/62234 to track the `TypeBuilder` problem.

Now I think it's probably better to not react to contextual reflection. If we did use contextual reflection, it would be possible to hit the `TypeBuilder` problem. If we always rely on the base type's ALC that will never happen.
Note that I don't think we will ever really fix the `TypeBuilder` problem, the best we can do without massive rewrite is to actually outright fail creating the type in the "failure" case. Which would mean that using contextual reflection would lead to cases where the proxy creating fails. Not great.

It would also mean that it would be possible to create two proxy types for the exact same combination of base type and interface type (in two different ALCs) - kind of against the current behavior. And there could also be subtle issues if the custom ALC specified via contextual reflection have different assembly resolution behavior.

It still possible to get to weird behavior cases, like loading the same `IFace` type into two different ALCs, and then trying to create proxies using the same `BaseType` but over the two different `IFace` types. But honestly I don't know if this has a solution - either way I can see this scenario to break, so picking one is probably as good as the other.

buyaa-n replied:
Thank you for looking into this

teo-tsirpanis committed:
Use Debug.Assert instead of an explicit check and throw.

jkotas commented on src/libraries/System.Reflection.DispatchProxy/src/System/Reflection/DispatchProxyGenerator.cs:
```diff
@@ -170,17 +91,34 @@ private sealed class GeneratedTypeInfo
-                _ab = AssemblyBuilder.DefineDynamicAssembly(new AssemblyName("ProxyBuilder"), AssemblyBuilderAccess.Run);
-                _mb = _ab.DefineDynamicModule("testmod");
+                string? alcName = alc.Name;
+                string name = string.IsNullOrEmpty(alcName) ? $"DispatchProxyTypes.{alc.GetHashCode()}" : $"DispatchProxyTypes.{alcName}";
```
> Storing the Default ALC's data in a separate static (outside of the weak table) - to avoid the cost of the weak table for the most common cases.

The weak table costs nothing compared to the cost of the DispatchProxy. I think it is fine to keep this detail simple and store everything in the same weak table.

buyaa-n replied:
Thank you for your review/suggestions @vitek-karas @jkotas, those are good points

@teo-tsirpanis as suggested above please use the old assembly name "ProxyBuilder" for the Default ALC  for back compatibility

teo-tsirpanis replied:
Done. I am wondering though, what is the risk of changing the dynamic assembly's name? How could user code depend on this?

vitek-karas replied:
We've learned the hard way not to assume anything. I agree it sounds weird, but if it's easy for us to maintain the same behavior, we always tend to do that.

teo-tsirpanis replied:
I see. As a precautionary measure, I renamed the dynamic module back to `testmod`.

teo-tsirpanis committed:
Address PR feedback.

Use a different name for the default ALC's proxy assembly and make it collectible only if its associated ALC is.

teo-tsirpanis commented on src/libraries/System.Reflection.DispatchProxy/src/System/Reflection/DispatchProxyGenerator.cs:
```diff
@@ -170,17 +91,34 @@ private sealed class GeneratedTypeInfo
-                _mb = _ab.DefineDynamicModule("testmod");
+                string? alcName = alc.Name;
+                string name = string.IsNullOrEmpty(alcName) ? $"DispatchProxyTypes.{alc.GetHashCode()}" : $"DispatchProxyTypes.{alcName}";
+                _ab = AssemblyBuilder.DefineDynamicAssembly(new AssemblyName(name), AssemblyBuilderAccess.RunAndCollect);
```
Done.

teo-tsirpanis committed:
Name the dynamic module "testmod" again.

buyaa-n commented:
/azp run runtime-libraries-coreclr outerloop

buyaa-n reviewed (approved):
Thank you @teo-tsirpanis, LGTM

buyaa-n commented:
Outerloop test failures are unrelated and reported as needed (https://github.com/dotnet/runtime/issues/62305,  https://github.com/dotnet/runtime/issues/60753#issuecomment-985084324, https://github.com/dotnet/runtime/issues/58616#issuecomment-985081412, https://github.com/dotnet/runtime/issues/56165#issuecomment-985088076). 

Thank you for your review @jkotas @vitek-karas if you have no more comments/concerns the PR is ready for merge.

vitek-karas commented:
Thanks a lot @teo-tsirpanis !!
